---
title: "CellTrails: Inference of Temporal Gene Expression Dynamics of Branching Biological Processes from Single-cell Expression Data"
shorttitle: "CellTrails: Single-cell Data Analysis"
author:
- name: "Daniel Christian Ellwanger"
  affiliation: "Department of Otolaryngology - Head & Neck Surgery, and Institute for Stem Cell Biology and Regenerative Medicine, Stanford University School of Medicine, Stanford, CA 94305, USA"
  email: "dc.ellwanger.dev@gmail.com"
- name: "Stefan Heller"
  affiliation: "Department of Otolaryngology - Head & Neck Surgery, and Institute for Stem Cell Biology and Regenerative Medicine, Stanford University School of Medicine, Stanford, CA 94305, USA"
package: "CellTrails"
tags: ["Trajectory reconstruction", "Dimensionality reduction", "Clustering", "Single-cell", "Machine learning"]

output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    fig_width: 4.5
    fig_height: 3.5
vignette: |
  %\VignetteIndexEntry{Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE) #, fig.width=4.5, fig.height=3.5, dpi=300 fig.align="left"
```

<!-- ---------------------------------- -->
# Prerequisites
<!-- ---------------------------------- -->
We strongly recommend to download and install the graph visualization software yEd (http://www.yworks.com/products/yed) [^1]. It provides great capabilities to perform planar embedding, visualization, and analysis of a trajectory graph produced by _CellTrails_. 

Before ready to use, the *CellTrails* libraries must be loaded into the _R_ environment:

```{r, eval=TRUE, message=FALSE, warning=FALSE}
library(CellTrails)
```

<!-- ---------------------------------- -->
# Detailed Tutorial
<!-- ---------------------------------- -->
**Each step of an typical _CellTrails_ analysis workflow is illustrated, described and explained in detail in this [tutorial](https://dcellwanger.github.io/CellTrails/).**
